### ISSUE TITLE
Integration with RayTune is currently somewhat broken (and so it the tutorial)

### ISSUE THREAD

**Issue by ilcord** [facebook/Ax#345]
I'm trying to set up a sobol+eb generation strategy and getting into some issues. Pretty sure I'm missing something very basic, so your help will be much appreciated.

Here's a snippet of what I'm trying to do:

```
# helpers for generation strategy creation
def _get_sobol_step(num_trials:int):
    return GenerationStep(
        model=Models.SOBOL,
        num_trials=num_trials,
        min_trials_observed=3,
        max_parallelism=None,
        enforce_num_trials=True,
        model_kwargs={"deduplicate": True, "seed": 123},
        index=0
    )

def _get_eb_step():
    return  GenerationStep(
            Models.EMPIRICAL_BAYES_THOMPSON, 
            num_trials=-1,
            min_trials_observed=0,
            max_parallelism=None, 
            enforce_num_trials=True,
            model_kwargs = {"transforms": TS_trans},
            index=1
        )

def gen_eb_strategy(init_trials:int = 5):    
    sobol, eb = _get_sobol_step(init_trials), _get_eb_step()
    return GenerationStrategy(steps=[sobol, eb], name='SOBOL+EB')


# create AxClient and experiment
eb_strategy = gen_eb_strategy(init_trials=5)
ax_client = AxClient(enforce_sequential_optimization=False, generation_strategy=eb_strategy)

ax_client.create_experiment(
    name="my_experiment",
    parameters=[
        {"name": "lr", "type": "choice", "values": [0.01, 0.02, 0.05], "value_type": "float"},
        {"name": "momentum", "type": "choice", "values": [0, 0.1, 0.2, 0.3, 0.5, 0.8], "value_type": "float"},
        # a lot of additional choice parameters
    ],
    objective_name="acc",
    overwrite_existing_experiment=True
)

ray.init(...)

# define train + evaluation method
def train_character_cnn(parameters: Dict[AnyStr, Any]):

    # training + evaluation logic here
    acc = ...

    track.log(
        acc=acc
    )

# run experiment
tune.run(
    train_character_cnn,
    num_samples=num_samples, 
    search_alg=AxSearch(ax_client),
    verbose=1,  # Set this level to 1 to see status updates and to 2 to also see trial results.
    resources_per_trial={"gpu": 0.25},
    local_dir='/tmp')
```

I'm getting this exception:
```
ValueError: StandardizeY transform requires non-empty observation data.
```

I'm not sure where I should plug in the observation data, and what data exactly I should add.

**Comment by eytan** [comment]
Somebody else on the team can answer your question about this exact error, but if you have many numerical parameters you should is there a reason you aren’t using the default EI searcher, which uses Bayesian optimization to efficiently infer what HP settings minimize your loss?  It will be substantially more efficient than any type of grid search—all you have to do is specify the parameters and their ranges (in general id recommend selecting learning rates on a log scale).

Empirical Bayes is only appropriate in the case where you have some pre defined discrete set of configurations (arms) with noisy observations.

Hope this helps
E

**Comment by ilcord** [comment]
Thanks for the clarification Eytan.

I did solve the issue at hand eventually, so I'll post the solution since I didn't find it anywhere in the documentation:

In my case of Sobol+EB (or the default Sobol+GP generation strategies), we have an initialization period (Sobol), followed by a more elaborate generation strategy that requires observation data (if I understand correctly, in the GP case, to model the prior).

To get the experiment to draw using both generation methods we first have to run `tune.run()` for enough samples (according to `min_trials_observed`), and then run `tune.run()` again (for as many samples we need, following best practices of max parallel draws, per batch). The second and following `tune.run()`s will be drawn from the GP generation step.

**Comment by lena-kashtelyan** [comment]
Hi, @ilcord, just wanted to follow up with you regarding this issue, although I think you already resolved it. I'll keep this issue open for now as we do need to look into why a single call to `tune.run` does not result in correct propagation of data to our models. Your solution sounds like a great workaround in the meantime!

It would also help us if you let us know why you opted for EB instead of EI, as @eytan asks above, just so we fully understand how you are looking to run your experiment.

**Comment by ilcord** [comment]
@lena-kashtelyan , just out of curiosity :) Pretty sure It'll come in handy at some point. In the end, I did end up using the default GPEI strategy.

I just want to clarify the issue, since I tried to reproduce it and found that it happens only with a slight modification to the code I posted:
When using a hybrid generation strategy (like, sobol + GPEI), the first call to `ray.run()` will always use the first generation step, no matter what will be the value of `num_samples`. To switch to the second one, I'll have to run `ray.run()` again.

If I'm using only GPEI, or only EB, I'll get the `ValueError: StandardizeY transform requires non-empty observation data.` error, which requires preliminary `ObservationData`.

Sorry for the inconsistency

**Comment by lena-kashtelyan** [comment]
@ilcord, gotcha, thank you!

**Comment by lena-kashtelyan** [comment]
The fix for this is now on master, and you can view the fixed tutorial on the 'latest' version of the website: https://ax.dev/versions/latest/tutorials/raytune_pytorch_cnn.html : ) This should also remove the need to call `ray.run()` twice, @ilcord.

**Comment by lena-kashtelyan** [comment]
The fix is now in the latest stable, 0.1.18.

**Comment by rogertrullo** [comment]
Hi, is this really solved? I am facing the same issues as @ilcord , I am doing something like this:

`step1=GenerationStep(Models.GPEI, num_trials=-1, min_trials_observed=0, max_parallelism=None, use_update=False, enforce_num_trials=True, model_kwargs=None, model_gen_kwargs=None, index=0)`

`strategy=GenerationStrategy([step1])`
`ax = AxClient(generation_strategy=strategy, enforce_sequential_optimization=False)`
which gives this strategy:
_GenerationStrategy(name='GPEI', steps=[GPEI for all trials])_
but then I run 
`tune.run(...)`
and I get this error:
_StandardizeY transform requires non-empty observation data._
